### ISSUE TITLE
Add benchmarks

### ISSUE THREAD

**Issue by jahtalab** [elastic/apm-agent-rum-js#26]
- [x] Micro benchmarks
     - [x] Add benchmarks for Transaction & Span creation overhead. 
- [x] Document benchmarks (Plus the Overhead)
- [x] Store benchmarks in ES
- [x] Create Kibana dashboards
- [x] Document the package size
- [x] Macro benchmarks on how the agent is affecting the user experience elastic/apm-agent-rum-js#302 
- [x] Run the benchmarks on Jenkins CI for publishing to ES
- [x] Run the benchmarks on bare metal server elastic/apm-agent-rum-js#456 

Postponed: 
- [ ] Run the benchmarks on each PR and add fields in the elasticsearch document - https://github.com/elastic/observability-dev/issues/299 -> https://github.com/elastic/apm-agent-rum-js/issues/568
- [x] Add memory benchmark for the browser - memory leaks -> https://github.com/elastic/apm-agent-rum-js/issues/514
- [ ] Consider moving the benchmark runner to the top to avoid reading/storing files multiple times

**Comment by jahtalab** [comment]
Preliminary results:

```
[
  {
    "name": "PerformanceMonitoring.createTransactionPayload",
    "mean": 0.000011806532270798866,
    "hz": 84698.87491632943
  },
  {
    "name": "PerformanceMonitoring.sendTransactions-no-json",
    "mean": 0.000023883519882611687,
    "hz": 41869.875333076285
  },
  {
    "name": "PerformanceMonitoring.sendTransactions",
    "mean": 0.00023317307692307694,
    "hz": 4288.659793814432
  }
]
```

**Comment by vigneshshanmugam** [comment]
Blocked on Jenkins Pipeline.

**Comment by jahtalab** [comment]
We still need to make the Kibana dashboards and also the scripts to send benchmark payload to ES. 
So it's currently not blocked.

**Comment by vigneshshanmugam** [comment]
Some thoughts on what are the dimensions that we have to come up on the RUM agent when it comes to the performance of the application and also the user experience. 

+ We have to document the different loading patterns and explain the pros/cons of the approach because this decides how our script parse and execution affects the page-load. We have to find how we **affect the browser from firing onload event**. 

+ **Parse time** of our agent script. Once the agent is downloaded, we patch some of the apis(XHR, Fetch) and also create global APM object. 

+ **Initialisation/Bootstrap** time, Basically when user calls `init` method, It sets up all the listeners, configs and other successive initialisation that needs for agent to function properly. 

+ **Payload Construction** - Once the load event is fired, most of our work comes up creating spans from RT and NT and also XHR spans that happened in the page. This is one of the most expensive work for the agent and we should identify the impact

+ **Payload Beacon** - Our payload is big right now since we do not have any compression technique, We have an open issue about this #158. We should also measure this as part of the process. 

+ **Post onload** - currently any transactions/errors that happens after onload is also captured and also for SPA, we have to measure how much impact performance by monitoring even beyond `onload`.

**Comment by vigneshshanmugam** [comment]
Automation team already has a script for publishing the benchmarks to the ES cluster when running on the Jenkins pipeline. We have to prepare the script for running the benchmarks on the ci

**Comment by vigneshshanmugam** [comment]
Closing this issue as we have opened dedicated ones for Performance Tuning and Benchmark son PR.
